app/post_metadata.go

```diff
@@ -0,0 +1,418 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import (
+	"image"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/dyatlov/go-opengraph/opengraph"
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/utils"
+	"github.com/mattermost/mattermost-server/utils/markdown"
+)
+
+const LINK_CACHE_SIZE = 10000
+const LINK_CACHE_DURATION = 3600
+
+var linkCache = utils.NewLru(LINK_CACHE_SIZE)
+
+func (a *App) InitPostMetadata() {
+	// Dump any cached links if the proxy settings have changed so image URLs can be updated
+	a.AddConfigListener(func(before, after *model.Config) {
+		if (before.ServiceSettings.ImageProxyType != after.ServiceSettings.ImageProxyType) ||
+			(before.ServiceSettings.ImageProxyURL != after.ServiceSettings.ImageProxyType) {
+			linkCache.Purge()
+		}
+	})
+}
+
+func (a *App) PreparePostListForClient(originalList *model.PostList) *model.PostList {
+	numPosts := len(originalList.Posts)
+
+	list := &model.PostList{
+		Posts: make(map[string]*model.Post, numPosts),
+		Order: originalList.Order, // Note that this uses the original Order array, so it isn't a deep copy
+	}
+
+	posts := make(chan *model.Post, numPosts)
+
+	for _, originalPost := range originalList.Posts {
+		go func(originalPost *model.Post) {
+			posts <- a.PreparePostForClient(originalPost)
+		}(originalPost)
+	}
+
+	for i := 0; i < numPosts; i++ {
+		post := <-posts
+		list.Posts[post.Id] = post
+	}
+
+	close(posts)
+
+	return list
+}
+
+func (a *App) PreparePostForClient(originalPost *model.Post) *model.Post {
+	post := originalPost.Clone()
+
+	// Proxy image links before constructing metadata so that requests go through the proxy
+	post = a.PostWithProxyAddedToImageURLs(post)
+
+	if post.Metadata == nil {
```

Comment:
This is left over from a previous version where I tried to only add any metadata fields that weren't already set so that we could prepopulate data if we already had it available. Obviously, we're not using that functionality though. Also, you're absolutely correct that it would double proxy the images anyway in that case.

That does get me thinking about where this count matter though. Now that I'm thinking about it, I'm not sure how this will behave when editing a post since it might send the old post up to the server which will already contain old metadata. I'll have to look into that

Reply:
I've changed it in https://github.com/mattermost/mattermost-server/pull/9868 so that it will always replace the metadata of the post when `PreparePostWithClient` is called since not doing that was making the tests problematic after some other changes to have `app.CreatePost` call `PreparePostWithClient` directly